test/matrix_test.go

```diff
@@ -0,0 +1,497 @@
+
+	taskRun4 := parse.MustParseV1TaskRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  params:
+  - name: GOARCH
+    value: linux/ppc64le
+  - name: package
+    value: path/to/common/package/
+  - name: version
+    value: go1.18.1
+  serviceAccountName: default
+  timeout: 1h0m0s
+  taskRef:
+    name: mytask
+    kind: Task
+status:
+  conditions:
+  - status: "True"
+    type: Succeeded
+`, "pr-matrix-include-3", namespace))
+
+	taskRun5 := parse.MustParseV1TaskRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  params:
+  - name: GOARCH
+    value: I-do-not-exist
+  serviceAccountName: default
+  timeout: 1h0m0s
+  taskRef:
+    name: mytask
+    kind: Task
+status:
+  conditions:
+  - status: "True"
+    type: Succeeded
+`, "pr-matrix-include-4", namespace))
+
+	taskrun1withresults := parse.MustParseV1TaskRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  results:
+  - name: GOARCHs
+    value:
+     - linux/amd64
+     - linux/ppc64le
+  serviceAccountName: default
+  timeout: 1h0m0s
+  taskRef:
+    name: task1withresults
+    kind: Task
+status:
+  conditions:
+  - status: "True"
+    type: Succeeded
+`, "pr-pt1-with-result", namespace))
+
+	taskrun2withresults := parse.MustParseV1TaskRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  results:
+  - name: versions
+    value:
+     - go1.17
+     - go1.18.1
+  serviceAccountName: default
+  timeout: 1h0m0s
+  taskRef:
+    name: task2withresults
+    kind: Task
+status:
+  conditions:
+  - status: "True"
+    type: Succeeded
+`, "pr-pt2-with-result", namespace))
+
+	expectedTaskRuns := []*v1.TaskRun{taskRun1, taskRun2, taskRun3, taskRun4, taskRun5, taskrun1withresults, taskrun2withresults}
+
+	t.Logf("Waiting for PipelineRun %s in namespace %s to complete", prName, namespace)
+	if err := WaitForPipelineRunState(ctx, c, prName, timeout, PipelineRunSucceed(prName), "PipelineRunSuccess", v1Version); err != nil {
+		t.Fatalf("Error waiting for PipelineRun %s to finish: %s", prName, err)
+	}
+
+	actualTaskrunList, err := c.V1TaskRunClient.List(ctx, metav1.ListOptions{LabelSelector: fmt.Sprintf("tekton.dev/pipelineRun=%s", prName)})
+	if err != nil {
+		t.Fatalf("Error listing TaskRuns for PipelineRun %s: %s", prName, err)
+	}
+
+	for i := range actualTaskrunList.Items {
+		expectedTaskRun := expectedTaskRuns[i]
+		if d := cmp.Diff(expectedTaskRun.Spec, actualTaskrunList.Items[i].Spec, ignoreTypeMeta); d != "" {
+			t.Errorf("expected to see TaskRun %v created. Diff %s", expectedTaskRuns[i].Name, diff.PrintWantGot(d))
+		}
+	}
+
+	t.Logf("Successfully finished test TestPipelineRunMatrixed")
+}
+
+// TestPipelineRunMatrixedFailed is an integration test with a Matrixed PipelineRun
+// that contains 2 successful, and 1 unsuccessful TaskRun. This test verifies that the
+// 3rd task run failing will cause the PipelineRun to fail, however the first 2 TaskRuns
+// should succeed before the last one fails.
```

Comment:
What guarantees the first 2 taskruns succeed before the third fails? Does this test have the potential to be flaky if the third fails too quickly?

Reply:
No, it'll just wait until the taskrun has reached the desired state, or the test times out. I don't think you can guarantee that parallel taskruns will finish in a certain order; you could add sleep statements but this can be brittle and adds latency. It's possible that the first test class you have is sufficient for e2e testing; do you think this test class covers cases that the first does not?

Reply:
Okay, I changed it just one taskRun checking the status that both the taskRun and PipelineRun failed. 
cc: @pritidesai

Reply:
If you want to continue with the original implementation, you can check the competition time of two succeeded taskRuns against the competition time of a failed task.